File: T-Bank/Algorithms-and-Data-Structures-2025/11sprint-Graph2/contest/E-Кратчайший_путь/solution.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <cstdint>



void solution() {
    /*
    Time Complexity O(NlogM)
    Memory Complexity O(N + M)
    */
    int n, m;
    std::scanf("%d %d", &n, &m);
    std::vector<bool> seen(n + 1, false);
    std::vector<long long> cost(n + 1, INT64_MAX);
    seen[1] = true;
    cost[1] = 0;
    std::vector<std::vector<std::pair<int, int>>> adjList(n + 1, std::vector<std::pair<int, int>>());
    for (int i = 0; i < m; ++i) {
        int source, destination, weight;
        std::scanf("%d %d %d", &source, &destination, &weight);
        adjList[source].push_back({destination, weight});
        adjList[destination].push_back({source, weight});
    }
    std::priority_queue<std::pair<long long, int>, std::vector<std::pair<long long, int>>, std::greater<std::pair<long long, int>>> minHeap;
    minHeap.push({0, 1});
    while (!minHeap.empty()) {
        int vertex = minHeap.top().second;
        long long amount = minHeap.top().first;
        minHeap.pop();
        for (std::pair<int, int> edge : adjList[vertex]) {
            if (!seen[edge.first] || (amount + edge.second < cost[edge.first])) {
                seen[edge.first] = true;
                cost[edge.first] = amount + edge.second;
                minHeap.push({amount + edge.second, edge.first});
            }
        }
    }
    for (int vertex = 1; vertex <= n; ++vertex) {
        std::printf("%lld ", cost[vertex]);
    }
    std::cout << '\n';
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/11sprint-Graph2/contest/E-Кратчайший_путь/solution.py
import sys
import heapq
from collections import defaultdict


def solution() -> None:
    """
    Time Complexity O(NlogM)
    Memory Complexity O(N + M)
    """
    n, m = map(int, sys.stdin.readline().rstrip().split())
    adj_list: dict[int, list[tuple[int, int]]] = defaultdict(list)
    seen: list[bool] = [False] * (n + 1)
    cost: list[int] = [float('inf')] * (n + 1)
    seen[1] = True
    cost[1] = 0
    for _ in range(m):
        source, destination, weight = map(int, sys.stdin.readline().rstrip().split())
        adj_list[source].append((destination, weight))
        adj_list[destination].append((source, weight))
    
    min_heap: list[tuple[int, int]] = [(0, 1)]
    while min_heap:
        amount, vertex = heapq.heappop(min_heap)
        for next_vertex, weight in adj_list[vertex]:
            if not seen[next_vertex] or (amount + weight < cost[next_vertex]):
                seen[next_vertex] = True
                cost[next_vertex] = amount + weight
                heapq.heappush(min_heap, (amount + weight, next_vertex))
    sys.stdout.write('{}\n'.format(' '.join(str(cost[vertex]) for vertex in range(1, n + 1))))


if __name__ == '__main__':
    solution()